src/univariate.jl

```diff
@@ -1,16 +1,50 @@
 ## Uniform ##
 
-logpdf(d::Uniform, x::TrackedReal) = uniformlogpdf(d.a, d.b, x)
-uniformlogpdf(a, b, x) = Tracker.track(uniformlogpdf, a, b, x)
+struct TuringUniform{T} <: ContinuousUnivariateDistribution
+    a::T
+    b::T
+end
+TuringUniform() = TuringUniform(0.0, 1.0)
+function TuringUniform(a::Int, b::Int)
+    return TuringUniform{Float64}(Float64(a), Float64(b))
+end
+function TuringUniform(a::Real, b::Real)
+    T = promote_type(typeof(a), typeof(b))
+    return TuringUniform{T}(T(a), T(b))
+end
+Distributions.logpdf(d::TuringUniform, x::Real) = uniformlogpdf(d.a, d.b, x)
+
+Distributions.Uniform(a::TrackedReal, b::Real) = TuringUniform{TrackedReal}(a, b)
+Distributions.Uniform(a::Real, b::TrackedReal) = TuringUniform{TrackedReal}(a, b)
+Distributions.Uniform(a::TrackedReal, b::TrackedReal) = TuringUniform{TrackedReal}(a, b)
+Distributions.logpdf(d::Uniform, x::TrackedReal) = uniformlogpdf(d.a, d.b, x)
+
+uniformlogpdf(a, b, x) = -log(b - a)
+uniformlogpdf(a::Real, b::Real, x::TrackedReal) = Tracker.track(uniformlogpdf, a, b, x)
+uniformlogpdf(a::TrackedReal, b::TrackedReal, x::Real) = Tracker.track(uniformlogpdf, a, b, x)
+uniformlogpdf(a::TrackedReal, b::TrackedReal, x::TrackedReal) = Tracker.track(uniformlogpdf, a, b, x)
 Tracker.@grad function uniformlogpdf(a, b, x)
-    xd = Tracker.data(x)
-    T = typeof(xd)
-    l = logpdf(Uniform(a, b), Tracker.data(x))
+    diff = Tracker.data(b) - Tracker.data(a)
+    T = typeof(diff)
+    l = -log(diff)
     f = isfinite(l)
-    da = 1/(b - a)
+    da = 1/diff
     n = T(NaN)
     return l, Δ->(f ? da : n, f ? -da : n, f ? zero(T) : n)
 end
+Zygote.@adjoint function uniformlogpdf(a, b, x)
```

Comment:
Zygote returns `nothing` for `x`'s pullback. When Tracker returned `nothing`, it was causing errors in Turing. So I fixed it then, and I applied the same fix for Zygote to make them consistent. There might be a better solution here. I can come back to it later.

Reply:
So in Turing we expect that the gradient is never `nothing` even if the output doesn't depend on the input. This is consistent with `ForwardDiff`. Because of how Tracker and Zygote work, the gradient is sometimes `nothing` if the output of the function is a constant. Perhaps we should consider the `nothing` case in Turing first then we can drop the workaround here. Let's do this in a separate PR.

Reply:
Hmm actually it's a bit more complicated. In Tracker, constant functions error https://github.com/FluxML/Tracker.jl/issues/48. Hence the need for a workaround here. The Zygote one is just to be consistent but is not necessary if `nothing` is handled properly in Turing.

Reply:
https://github.com/TuringLang/Turing.jl/issues/1081

Reply:
Isn't the (main) problem with Tracker that the output types should depend on the input (types), even if the output is constant? E.g., `Tracker.gradient(x -> one(eltype(x)), [0.5])` (just took the example from the Tracker issue that you linked) returns the expected gradient of 0.

Reply:
Yes this would have to be fixed in Distributions then.